castor-service/src/main/java/io/carbynestack/castor/service/persistence/cache/CreateReservationSupplier.java

```diff
@@ -51,55 +48,75 @@ public Reservation get() {
-    long stillToReserve = numberOfTuples;
-    while (stillToReserve > 0) {
-      try {
-        TupleChunkFragmentEntity availableFragment =
-            fragmentStorageService
-                .findAvailableFragmentWithTupleType(tupleType)
-                .orElseThrow(
-                    () ->
-                        new CastorServiceException(FAILED_FETCH_AVAILABLE_FRAGMENT_EXCEPTION_MSG));
-        long tuplesInFragment = availableFragment.getEndIndex() - availableFragment.getStartIndex();
-        if (tuplesInFragment > stillToReserve) {
-          availableFragment =
-              fragmentStorageService.splitAt(
-                  availableFragment, availableFragment.getStartIndex() + stillToReserve);
-        }
-        availableFragment.setReservationId(reservationId);
-        fragmentStorageService.update(availableFragment);
-        long tuplesTaken = Math.min(tuplesInFragment, stillToReserve);
-        stillToReserve -= tuplesTaken;
-        reservationElements.add(
-            new ReservationElement(
-                availableFragment.getTupleChunkId(),
-                tuplesTaken,
-                availableFragment.getStartIndex()));
-      } catch (Exception e) {
-        throw new CastorServiceException(FAILED_RESERVE_TUPLES_EXCEPTION_MSG, e);
+    long oddToReserve = numberOfTuples % castorServiceProperties.getInitialFragmentSize();
+    long roundToReserve = numberOfTuples - oddToReserve;
+    List<ReservationElement> roundElements = null;
+    if (roundToReserve > 0) {
+      ArrayList<TupleChunkFragmentEntity> roundFragments =
+          fragmentStorageService.retrieveAndReserveRoundFragments(
+              (int) roundToReserve, tupleType, reservationId);
+
+      // if the fragments are split up there might be enough tuples in the non-round fragments.
+      // -- especially with inputmasks
+      if (roundFragments.size() * (long) castorServiceProperties.getInitialFragmentSize()
+          < roundToReserve)
+        oddToReserve +=
+            roundToReserve
+                - roundFragments.size() * (long) castorServiceProperties.getInitialFragmentSize();
+
+      // maps all fragments returned to ReservationElements using their 'tupleChunnkId' attribute,
+      // the common fragment size and their 'startIndex' attribute
+      roundElements =
+          roundFragments.stream()
+              .map(
+                  frag ->
+                      new ReservationElement(
+                          frag.getTupleChunkId(),
+                          castorServiceProperties.getInitialFragmentSize(),
+                          frag.getStartIndex()))
+              .collect(Collectors.toList());
+    }
+    while (oddToReserve > 0) {
+      TupleChunkFragmentEntity availableFragment =
+          fragmentStorageService
+              .retrieveSinglePartialFragment(
+                  tupleType, oddToReserve < castorServiceProperties.getInitialFragmentSize())
```

Comment:
> second argument `preferSmall` of `retrieveSinglePartialFragment(TupleType tupleType, boolean preferSmall)` is always true, does this make sense?

This argument is not always true. `oddToReserve` may be greater than the initial Fragment size, if Castor couldn't get the whole amount of Tuples in whole fragments. This happened in my test setup because Klyshko had an off-by-one error and would save 99,999 Tuples instead of 100,000 Tuples into a `tupleChunk` resulting in lots of tuples getting saved in partial fragments.

Line 79 adds the amount of tuples to `oddToReserve` that _should_ have been reserved as round, couldn't.